src/stirling/source_connectors/perf_profiler/pprof/pprof.cc

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "src/stirling/source_connectors/perf_profiler/pprof/pprof.h"
+
+#include <gflags/gflags.h>
+
+#include <absl/strings/str_split.h>
+#include <vector>
+
+DECLARE_uint32(stirling_profiler_stack_trace_sample_period_ms);
+
+namespace px {
+namespace stirling {
+
+PProfProfile CreatePProfProfile(const uint32_t num_cpus, const histo_t& histo) {
+  // Info on the pprof proto format:
+  // https://github.com/google/pprof/blob/main/proto/profile.proto
+
+  // period_ns will be used when populating the nanos count.
+  const uint64_t period_ms = FLAGS_stirling_profiler_stack_trace_sample_period_ms;
+  const uint64_t period_ns = period_ms * 1000 * 1000;
+
+  // Tracks which strings have been inserted into the profile.
+  absl::flat_hash_map<std::string, uint64_t> strings;
+
+  // This is the pprof profile.
+  ::perftools::profiles::Profile profile;
+
+  // The pprof profiles start by describing their sample types. For CPU profiling,
+  // the convention is to describe two different metrics:
+  // "samples" with units of "count", and
+  // "cpu" with units of "nanoseconds".
+  // Note, the first entry in the strings table is required to be an empty string.
+  auto sample_type = profile.add_sample_type();
+  sample_type->set_type(1);
+  sample_type->set_unit(2);
+  profile.add_string_table("");
+  profile.add_string_table("samples");
+  profile.add_string_table("count");
+  sample_type = profile.add_sample_type();
+  sample_type->set_type(3);
+  sample_type->set_unit(4);
+  profile.add_string_table("cpu");
+  profile.add_string_table("nanoseconds");
+
+  // State variables useful as we build the pprof profile message.
+  // No locations messages exist (yet); next_location_id starts at 1.
+  // We already added some strings; next_string_id starts at 5.
+  uint64_t next_location_id = 1;
+  uint64_t next_string_id = 5;
+
+  // To build the profile, we iterate over the stack traces  histogram.
```

Comment:
>   // To build the profile, we iterate over the stack traces histogram.

There is an extra space between "traces" and "histogram".

Reply:
Yikes, thanks for noticing!